libs/simdvec/src/main21/java/org/elasticsearch/simdvec/internal/vectorization/PanamaESVectorUtilSupport.java

```diff
@@ -83,6 +106,267 @@ public float ipFloatByte(float[] q, byte[] d) {
         return DefaultESVectorUtilSupport.ipFloatByteImpl(q, d);
     }
 
+    @Override
+    public void centerAndCalculateOSQStatsEuclidean(float[] vector, float[] centroid, float[] centered, float[] stats) {
+        assert vector.length == centroid.length;
+        assert vector.length == centered.length;
+        float vecMean = 0;
+        float vecVar = 0;
+        float norm2 = 0;
+        float min = Float.MAX_VALUE;
+        float max = -Float.MAX_VALUE;
+        int i = 0;
+        int vectCount = 0;
+        if (vector.length > 2 * FLOAT_SPECIES.length()) {
+            FloatVector vecMeanVec = FloatVector.zero(FLOAT_SPECIES);
+            FloatVector m2Vec = FloatVector.zero(FLOAT_SPECIES);
+            FloatVector norm2Vec = FloatVector.zero(FLOAT_SPECIES);
+            FloatVector minVec = FloatVector.broadcast(FLOAT_SPECIES, Float.MAX_VALUE);
+            FloatVector maxVec = FloatVector.broadcast(FLOAT_SPECIES, -Float.MAX_VALUE);
+            int count = 0;
+            for (; i < FLOAT_SPECIES.loopBound(vector.length); i += FLOAT_SPECIES.length()) {
+                ++count;
+                FloatVector v = FloatVector.fromArray(FLOAT_SPECIES, vector, i);
+                FloatVector c = FloatVector.fromArray(FLOAT_SPECIES, centroid, i);
+                FloatVector centeredVec = v.sub(c);
+                FloatVector deltaVec = centeredVec.sub(vecMeanVec);
+                norm2Vec = fma(centeredVec, centeredVec, norm2Vec);
+                vecMeanVec = vecMeanVec.add(deltaVec.div(count));
+                FloatVector delta2Vec = centeredVec.sub(vecMeanVec);
+                m2Vec = fma(deltaVec, delta2Vec, m2Vec);
+                minVec = minVec.min(centeredVec);
+                maxVec = maxVec.max(centeredVec);
+                centeredVec.intoArray(centered, i);
+            }
+            min = minVec.reduceLanes(MIN);
+            max = maxVec.reduceLanes(MAX);
+            norm2 = norm2Vec.reduceLanes(ADD);
+            vecMean = vecMeanVec.reduceLanes(ADD) / FLOAT_SPECIES.length();
+            FloatVector d2Mean = vecMeanVec.sub(vecMean);
+            m2Vec = fma(d2Mean, d2Mean, m2Vec);
+            vectCount = count * FLOAT_SPECIES.length();
+            vecVar = m2Vec.reduceLanes(ADD);
+        }
+
+        float tailMean = 0;
+        float tailM2 = 0;
+        int tailCount = 0;
+        // handle the tail
+        for (; i < vector.length; i++) {
+            centered[i] = vector[i] - centroid[i];
+            float delta = centered[i] - tailMean;
+            ++tailCount;
+            tailMean += delta / tailCount;
+            float delta2 = centered[i] - tailMean;
+            tailM2 = fma(delta, delta2, tailM2);
+            min = Math.min(min, centered[i]);
+            max = Math.max(max, centered[i]);
+            norm2 = fma(centered[i], centered[i], norm2);
+        }
```

Comment:
@tveasey could you take a look here? I think I did the variance calculation here correctly. But I might have missed something.

Reply:
Formulas look correct to me of course checking with vector lengths 10 - 100 vs slow calculation is within a few epsilon will prove there are no errors. Specifically, I would add testing of vectorised stats calculation against the super simple vestions, i.e. compute mean, then compute mean of square residuals, etc, so there is no chance of errors. If you're v.close to that a bunch of random length vectors you're good. (This may be what the Lucene reference does, but if it using online calculation my inclination would be to simplify further so there is no chance of errors.)